[RUM-9902] Isolating implementation layer from integration modules

satween opened the pull request:
### What does this PR do?

The goal of this PR is to isolate the implementation from the API specification. The Datadog tracing integration is provided by the dd-trace-java library, which we moved into the dd-android-trace-internal module in a previous change.

We are now trying to finally eliminate `opentracing` dependencies and are facing the following issues:
1. The `opentracing`  API specification is used as an internal layer for module communication. For example, the `dd-android-sdk-trace-http` module uses the following objects from `opentracing`: `Span`, `Scope`, etc. The diagram below displays the existing architecture:
 
```mermaid
graph BT
J(otel) --> H(io.opentelemetry)
B(trace) --> L(trace-internal)
M[ ]:::empty --> B
J --> L
F(trace-coroutines) --- M
B -- api --> A(io.opentracing)
E(sqldelight) --- M
D(okhttp) --- M
D -- api--> A
classDef empty width:0px,height:0px;
```
_As you might have noticed, there are a lot of modules that are aware of the specification classes and are bound to a specific implementation of the `opentracing` specification, such as `AndroidTracing`._

2. The second problem is that most modules know not only the specifications, but also the implementation.
A lot of downcasting logic can be found inside the integration module:
```kotlin
if (activeSpan is DDSpan) {
        activeSpan.traceId.toString()
    }
```

To solve the described issues and decouple the implementation from the internal and external specifications, we provide the following architecture:

```mermaid
graph BT
C --impl--> B(trace-internal)
C(trace) --api--> A(trace-api)
C --(ready to remove)--> O(io.opentracing)
M[ ]:::empty --> C
D(okhttp)  --- M
F(trace-coroutines) --- M
E(sqldelight) --- M
J(otel) --- M
J --> H(io.opentelemetry) 
classDef empty width:0px,height:0px;
```

This PR mainly includes:
1. The `dd-sdk-android-trace-api` module was added to provide the internal API specification for trace functionality.
2. The `dd-android-sdk-trace` module is now the only point that knows both the specification and the implementation.
3. The `opentracing` API specification was replaced with the `dd-sdk-android-trace-api` for all modules that used the `opentracing` specification previously.
4.All `api` dependencies were replaced with `implementation` dependencies for all modules except for `dd-sdk-android-trace-api`.


TL;DR:
In this pull request, we isolated the specific implementation from the specification and removed all of the downcasting logic from the integration modules. Now, the only place where the implementation meets the specification is the dd-sdk-android-trace module.

### Motivation 

> Why just not to use Otel's specification instead of opentraicing? 

 Otel requires API 26 or desugaring support

> Why just not to use `dd-sdk-android-trace-internal`'s interfaces as a specification? 

Most of the files in the `dd-sdk-android-trace-internal` directory are written in Java because it adopts the `dd-trace-java` directory. Therefore, some Kotlin features, such as the internal modifier, will not work if they are made public. Additionally, `dd-trace-java` was initially designed for a server-side agent and contains specific methods that are not applicable to Android, which could confuse customers.

### Additional Notes

Note that there is still dependency on opentracing and internal module mostly for tests. I decided to remove them in a separate PR to reduce amount of changes. 


### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [ ] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [ ] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)

satween committed:
RUM-9902: adding trace-api module

satween committed:
RUM-9902: decoupling internal from traces

satween committed:
RUM-9902: decoupling internal from traces

satween committed:
RUM-9902: decoupling internal from traces (200 tests fallen)

satween committed:
RUM-9902: decoupling internal from traces (103 tests fallen)

satween committed:
RUM-9902: decoupling internal from traces (31 tests fallen)

satween committed:
RUM-9902: decoupling internal from traces (24 tests fallen)

satween committed:
RUM-9902: decoupling internal from traces (10 tests fallen)

satween committed:
RUM-9902: decoupling internal from traces (3 tests fallen)

satween committed:
RUM-9902: decoupling internal from traces (0 tests fallen)

satween committed:
RUM-9902: decoupling internal from traces (broken integration tests)

satween committed:
RUM-9902: decoupling internal from traces

satween committed:
RUM-9902: decoupling internal from traces

satween committed:
RUM-9902: decoupling internal from traces

satween committed:
RUM-9902: decoupling internal from traces - removing opentracing dependencies from wear app

satween committed:
RUM-9902: decoupling internal from traces - NoOp support for tracer

satween committed:
RUM-9902: decoupling internal from traces - Fix tests

satween committed:
RUM-9902: decoupling internal from traces - Initialization in wear

satween committed:
RUM-9902: decoupling internal from traces - fix logs. fail by warn and tracing builder

satween committed:
RUM-9902: decoupling internal from traces - getting rid from Adroid traicer depeendencies

satween committed:
RUM-9902: decoupling internal from traces - reverting OtelTracerBuilderProviderTest

satween committed:
RUM-9902: decoupling internal from traces - revert deprecation suppression

satween committed:
RUM-9902: decoupling internal from traces - revert deprecation suppression

satween committed:
RUM-9902: decoupling internal from traces - fix tests

satween committed:
RUM-9902: decoupling internal from traces - fix tests

satween committed:
RUM-9902: decoupling internal from traces - fix tests

satween committed:
RUM-9902: decoupling internal from traces - fix detekt

satween committed:
RUM-9902: pre-review fixes - removing dependency to -trace-internal lib

satween committed:
RUM-9902: pre-review fixes - reducing changes amount

satween committed:
RUM-9902: pre-review fixes - reducing changes amount

satween committed:
RUM-9902: pre-review fixes - reducing changes amount

satween committed:
RUM-9902: pre-review fixes - reducing changes amount

satween committed:
RUM-9902: pre-review fixes - reducing changes amount

satween committed:
RUM-9902 - merge develop

satween committed:
RUM-9902 - merge develop

mariusc83 reviewed (commented):
The architecture looks clean but what I am worried about is all these adapters and wrappers which are being used in `Otel` module now. Was it really needed ? This module deviates a lot now from the one in the APM java Otel module so it is going to be hard to keep it up-to-date.

satween committed:
RUM-9902 - fix transitiveDependencies

satween commented on features/dd-sdk-android-trace/src/main/kotlin/com/datadog/android/trace/GlobalDatadogTracerHolder.kt:
```diff
@@ -0,0 +1,44 @@
+ */
+object GlobalDatadogTracerHolder {
+    @get:Synchronized
+    internal var tracer: DatadogTracer? = null
```
Good point 👍

satween committed:
RUM-9902 - support volatile

0xnm commented on integrations/dd-sdk-android-okhttp/src/main/kotlin/com/datadog/android/okhttp/trace/TracingInterceptor.kt:
```diff
@@ -521,11 +524,11 @@ internal constructor(
         isSampled: Boolean
     ): Request.Builder {
-        val tracedRequestBuilder = request.newBuilder()
+        val tracedRequestBuilder: Request.Builder = request.newBuilder()
```
what is the need for the explicit type?

satween replied:
debug leftover

0xnm reviewed (commented):
💪 

I went through the production code, not spending much time on the tests for now and left some comments.

satween committed:
RUM-9902 - fix tests

satween committed:
RUM-9902 - fix tests

satween committed:
RUM-9902 - fix tests, getting rid from TraceContext

satween committed:
RUM-9902 - fix tests, getting rid from TraceContext

satween committed:
RUM-9902 - getting rid from tracer context

satween committed:
RUM-9902 - getting rid from properties in OtelTracerProviderBuilder

satween committed:
RUM-9902 - changing values in tests:  DATADOG_SAMPLING_PRIORITY_HEADER set to 2 instead of 1 because RuleBasedTraceSampler sets USER_DROP/USER_KEEP instead of SAMPLE_DROP/SAMPLE_KEEP

satween committed:
RUM-9902 - changing values in tests:  changing order of event written, because  new implementation (Coretracer) works bit different that opentracing.
In the previous implementation TracerInterceptor used AndroidTracer as an engine for spans writing and DEFAULT_PARTIAL_MIN_FLUSH = 5 parameter was ignored.
New version of TracingInterceptor using CoreTracer under the hood, and DEFAULT_PARTIAL_MIN_FLUSH is taken into account.
Note that reducing DEFAULT_PARTIAL_MIN_FLUSH to 1 will give same result as AndroidTracing.

satween committed:
RUM-9902 - moving tests from OtelTracerBuilderProviderTest into DatadogTracingTest and fix others

satween committed:
Apply suggestions from code review

Co-authored-by: Nikita Ogorodnikov <4046447+0xnm@users.noreply.github.com>

satween committed:
RUM-9902 - pos-review fixes: adding more details for module description

satween committed:
Merge remote-tracking branch 'origin/tvaleev/feature/RUM-9902-removing-unused-code' into tvaleev/feature/RUM-9902-removing-unused-code

satween committed:
RUM-9902 - pos-review fixes: removing unnecessary packages

satween commented on features/dd-sdk-android-trace-internal/build.gradle.kts:
```diff
@@ -54,7 +54,7 @@ dependencies {
 
     // OpenTracing
-    api(libs.bundles.openTracing)
+    implementation(libs.bundles.openTracing)
```
yes. keeping it for now because amount of changes is big enough for one PR

satween commented on features/dd-sdk-android-trace/src/main/kotlin/com/datadog/android/trace/SpanExt.kt:
```diff
@@ -6,8 +6,9 @@
 
+import com.datadog.android.trace.api.span.DatadogSpan
+import com.datadog.android.trace.impl.DatadogTracing
 import io.opentracing.Span
```
gonna remove with io.opentracing packages removing in a separate PR

satween commented on features/dd-sdk-android-trace/src/testFixtures/kotlin/com/datadog/android/trace/impl/TracingErrorMessages.kt:
```diff
@@ -0,0 +1,14 @@
+
+package com.datadog.android.trace.impl
+
+object TracingErrorMessages {
```
it's a bridge for error messages used only in tests and added as a `testFixtures`, so I think to have it public

satween committed:
RUM-9902 - pos-review fixes

satween committed:
RUM-10719 - removing opentracing from dd-sdk-android-trace-internal module

satween committed:
RUM-10719 - removing opentracing from dd-sdk-android-trace module

satween committed:
RUM-10719 - removing opentracing from dd-sdk-android-rum  module

satween committed:
RUM-10719 - removing opentracing from dd-sdk-android-reliability-single-fit  module

satween committed:
RUM-10719 - removing opentracing from instrumented-integration  module

satween committed:
RUM-10719 - fix tests

satween committed:
RUM-10719 - fix static analysis

satween committed:
RUM-10719 - restoring removed test

satween committed:
RUM-10719 - restoring removed test

satween committed:
RUM-10719 - fix ktlint

satween committed:
RUM-10719 - fix detekt

satween committed:
RUM-10719 - getting rid from legacy package

satween committed:
RUM-10719 - fix test

satween committed:
RUM-10719 - fix test

satween committed:
RUM-10719 - rolling back StubFeatureScope to avoid OOMs

satween committed:
Merge pull request #2783 from DataDog/tvaleev/feature/RUM-10719-removing_opentracing

[RUM-10719]: Removing opentracing dependencies

satween commented on features/dd-sdk-android-trace-otel/src/main/kotlin/com/datadog/android/trace/opentelemetry/OtelTracerProvider.kt:
```diff
@@ -144,54 +145,57 @@ class OtelTracerProvider internal constructor(
-                    ?.unwrap<Feature>()
+                val tracingFeature = sdkCore.getFeature(Feature.TRACING_FEATURE_NAME)?.unwrap<Feature>()
                 val internalCoreWriterProvider = tracingFeature as? InternalCoreWriterProvider
-                if (tracingFeature == null) {
```
This logic was moved into `DatadogTracing` object to have same behavior when `DatadogTracer` instance is being created no matter if it's otel or direct creation

satween commented on features/dd-sdk-android-trace/src/main/kotlin/com/datadog/android/trace/impl/internal/DatadogSpanLinkAdapter.kt:
```diff
@@ -0,0 +1,32 @@
+import com.datadog.trace.bootstrap.instrumentation.api.SpanLink
+import com.datadog.trace.bootstrap.instrumentation.api.SpanLinkAttributes
+
+internal class DatadogSpanLinkAdapter(delegate: DatadogSpanLink) :
```
First, `SpanLink` is a data class. There are no setters, the constructor is protected, etc.

Second, yes, you are right in most cases, the Adapter class wraps the internal implementation and delegates everything to it. However, with `SpanLink`, it's a bit different. The OpenTelemetry specification declares a method with the following signature:

```java
 SpanBuilder addLink(SpanContext spanContext, Attributes attributes);
```

The class Attributes is part of the OpenTelemetry specification, so we cannot simply pass it into the "-trace-api" or "-trace-internal" modules.

On the other hand, we don't want to expose the `SpanLink` class of the `-trace-internal` module. The only option is to wrap it in an adapter.

Next, if we want to wrap it in an adapter in the `-otel` module, we need to either make the `DatadogSpanLinkAdapter` class public or create a builder method inside the `DatadogTracing` object. Creating a method for a specific `SpanLink` scenario would make the SDK harder for most customers to understand, but only a few customers (those who use SpanLink) would benefit from it.

On the other hand, the current implementation lets us:
- Keep the `DatadogSpanLinkAdapter` class internal, meaning the `-otel` module doesn't know any internal implementation details.
- Keep the `DatadogTracing` object with just a single public method (that lets customers create a `DatadogTracer` object), which makes the SDK easier for customers.

satween committed:
RUM-9902 - simplifying api, getting rid from wrapper classes

satween committed:
RUM-9902 - simplifying api

satween commented on features/dd-sdk-android-trace-api/src/main/kotlin/com/datadog/android/trace/api/span/DatadogSpanContext.kt:
```diff
@@ -0,0 +1,58 @@
+     */
+    fun setSamplingPriority(samplingPriority: Int): Boolean
+
+    /**
+     * Sets the tracing sampling priority if it is necessary.
+     */
+    @InternalApi
+    fun setTracingSamplingPriorityIfNecessary()
```
yep, hidden,  it's also leftover for migration purposes.

satween commented on features/dd-sdk-android-trace-api/src/main/kotlin/com/datadog/android/trace/api/span/DatadogSpanIdConverter.kt:
```diff
@@ -0,0 +1,31 @@
+     * @param spanId The span ID to convert, represented as a long value.
+     * @return A hexadecimal string representation of the given span ID, padded as necessary.
+     */
+    fun toHexStringPadded(spanId: Long): String
```
Well, original method(in DDSpanId) doesn't say how it's padded as well. we basically using this api only for internal purposes (Otel integration + tests) so I think is okay to keep it as it is

satween commented on features/dd-sdk-android-trace-api/src/main/kotlin/com/datadog/android/trace/api/span/DatadogSpanWriter.kt:
```diff
@@ -0,0 +1,9 @@
+package com.datadog.android.trace.api.span
+
+/** A writer is responsible to send collected spans to some place. */
+interface DatadogSpanWriter
```
Unfortunately we need this interface only to wrap the `com.datadog.trace.common.writer.Writer` class otherwise it's impossible to keep `com.datadog.trace.common.writer.Writer` inside `trace-internal` module, because `InternalCoreWriterProvider` is a part of `-trace` module and public to descendants.

satween commented on features/dd-sdk-android-trace-api/src/main/kotlin/com/datadog/android/trace/api/tracer/DatadogTracerBuilder.kt:
```diff
@@ -0,0 +1,79 @@
+     * @param traceId128BitGenerationEnabled A flag indicating whether 128-bit trace ID generation should be enabled.
+     * @return The updated instance of [DatadogTracerBuilder] to allow method chaining.
+     */
+    fun withIdGenerationStrategy(key: String, traceId128BitGenerationEnabled: Boolean): DatadogTracerBuilder
```
hidden

satween commented on features/dd-sdk-android-trace-otel/src/main/java/com/datadog/opentelemetry/trace/OtelSpanLink.java:
```diff
@@ -6,67 +6,120 @@
-    if (attributes.isEmpty()) {
-      return SpanLinkAttributes.EMPTY;
+  private static <T> void putArray(Map<String, String> attributes, String key, List<T> array) {
+    requireNonNull(key, "key must not be null");
```
fixed

satween committed:
RUM-9902 - simplifying api

satween commented on features/dd-sdk-android-trace-otel/src/main/java/com/datadog/opentelemetry/trace/OtelSpan.java:
```diff
@@ -6,18 +6,24 @@
 package com.datadog.opentelemetry.trace;
 
-import static com.datadog.trace.api.ConfigDefaults.DEFAULT_ASYNC_PROPAGATING;
+import static com.datadog.android.trace.api.constants.DatadogTracingConstants.DEFAULT_ASYNC_PROPAGATING;
```
yep, removed

satween commented on features/dd-sdk-android-trace-api/src/main/kotlin/com/datadog/android/trace/api/trace/DatadogTraceIdFactory.kt:
yep, moved into `-trace` module along with `DatadogSpanIdConverter`

satween commented on features/dd-sdk-android-trace-api/src/main/kotlin/com/datadog/android/trace/api/tracer/DatadogTracerSampler.kt:
```diff
@@ -0,0 +1,15 @@
+ * a span should be traced.
+ */
+@NoOpImplementation
+interface DatadogTracerSampler
```
this whole part is removed, figured out that it's not required to make it public

satween committed:
RUM-9902: post review fixes

satween committed:
RUM-9903 - upgrading opentelemtry version, removing separate benchmark version

RUM-9902 - fix detekt and tests

0xnm commented on features/dd-sdk-android-trace-api/src/main/kotlin/com/datadog/android/trace/api/scope/DataScopeListener.java:
```diff
@@ -9,9 +9,7 @@
   /**
-   * Called just after a scope becomes the active scope
-   * May be called multiple times. When a scope is initially created, or after a child scope is
-   * deactivated.
```
it cannot be called multiple times anymore?

satween replied:
Looks like lost this changes in one of updates, restored

0xnm commented on features/dd-sdk-android-trace-api/src/main/kotlin/com/datadog/android/trace/api/span/DatadogSpan.kt:
```diff
@@ -64,7 +64,7 @@ interface DatadogSpan {
      * Represents the start time of the span in nanoseconds.
      */
-    val startTime: Long
+    val startTimeNano: Long
```
just for the consistency with naming across the codebase

```suggestion
    val startTimeNanos: Long
```

0xnm commented on features/dd-sdk-android-trace-api/src/main/kotlin/com/datadog/android/trace/api/span/DatadogSpanBuilder.kt:
```diff
@@ -34,8 +33,6 @@ interface DatadogSpanBuilder {
 
     /**
      * Adds a tag to the span with the specified key and a nullable double value.
-     * Tags are used to add contextual information to spans, which can later be
```
why we are deleting this? not useful?

satween replied:
Seems like being lost during one of squash/rebase procedure

0xnm commented on features/dd-sdk-android-trace-api/src/main/kotlin/com/datadog/android/trace/api/trace/DatadogTraceId.kt:
```diff
@@ -13,18 +13,4 @@ import com.datadog.tools.annotation.NoOpImplementation
-     */
-    fun toHexString(): String
-}
+interface DatadogTraceId
```
is it useful to have empty interface?

satween replied:
`DatadogTraceId ` - well, it's was many approaches of how to use traceId. In the current dd-core-java implementation we using in our SDK TraceId is an object that basically consist of two long values. So mainly all methods it has is just convertation from/to Long or String value. 

Because we use `DatadogTraceId` as middle layer between SDK code and dd-trace-java it's impossible to make them static (`DatadogTraceId` is part of API, and dd-trace-java's DDTraceId - an implementation).  So first I was thinking to put whole logic related to conversion into `DatadogTraceIdConverter`. but then we end up in a situation when we have empty interfaces. 

After analyzing the codebase and how those conversion methods being used I realized that in most cases they required for test code only. So I put everything related to production code into `DatadogTraceId ` interface. 

Note that  there is empty companion object there required to be able to add some kotlin sugar and mimic extension method as an static method: 

```kotlin

fun DatadogTraceId.Companion.fromHex(traceId: String): DatadogTraceId {
    return DatadogTraceIdAdapter(DDTraceId.fromHex(traceId))
}

```

let then use it like: 

```kotlin
DatadogTraceId.fromHex(...)
```

0xnm commented on features/dd-sdk-android-trace/src/main/kotlin/com/datadog/android/trace/impl/DatadogTracing.kt:
```diff
@@ -5,42 +5,29 @@
 import com.datadog.android.trace.impl.internal.DatadogTracerBuilderAdapter
-import com.datadog.android.trace.internal.data.NoOpCoreTracerWriter
+import com.datadog.android.trace.impl.internal.DatadogTracingInternalToolkit
+import com.datadog.android.trace.impl.internal.DatadogTracingInternalToolkit.ErrorMessages.DEFAULT_SERVICE_NAME_IS_MISSING_ERROR_MESSAGE
+import com.datadog.android.trace.impl.internal.DatadogTracingInternalToolkit.ErrorMessages.TRACING_NOT_ENABLED_ERROR_MESSAGE
+import com.datadog.android.trace.impl.internal.DatadogTracingInternalToolkit.ErrorMessages.WRITER_PROVIDER_INTERFACE_NOT_IMPLEMENTED_ERROR_MESSAGE
```
do we want to share error messages? is it useful?

satween replied:
as discussed, moved into `DatadogTracing`

0xnm commented on features/dd-sdk-android-trace/src/main/kotlin/com/datadog/android/trace/impl/DatadogTracing.kt:
```diff
@@ -52,62 +39,49 @@ object DatadogTracing {
+            DatadogTracingInternalToolkit.builderProvider as DatadogTracerBuilder
+        }
+        sdkCore !is FeatureSdkCore -> {
+            NoOpDatadogTracerBuilder()
```
maybe we need to log a warning here?

satween replied:
if `sdkCore` is not an `FeatureSdkCore` we don't have an access to `internalLogger`

0xnm commented on features/dd-sdk-android-trace/src/main/kotlin/com/datadog/android/trace/impl/internal/DatadogScopeListenerAdapter.kt:
```diff
@@ -5,10 +5,13 @@
 
-internal class DatadogScopeListenerAdapter(private val delegate: DataScopeListener) : ScopeListener {
+internal class DatadogScopeListenerAdapter(
+    @get:VisibleForTesting internal val delegate: DataScopeListener
```
minor: not sure if `VisibleForTesting` annotation is useful for Kotlin code. It is different for Java, where it is usually applied to the `public` members and is usually and indicator that such member still not a part of official API, but here it is internal anyway.

satween replied:
I put `@VisibleForTesting` annotation in such places as an explanation why this property should be `internal` instead of private.  Like here it's means that it's internal only for the testing purposes.

0xnm commented on features/dd-sdk-android-trace/src/main/kotlin/com/datadog/android/trace/impl/internal/DatadogSpanAdapter.kt:
```diff
@@ -22,7 +22,7 @@ internal class DatadogSpanAdapter(internal val delegate: AgentSpan) : DatadogSpa
     override val durationNano: Long get() = delegate.durationNano
 
-    override val startTime: Long get() = delegate.startTime
+    override val startTimeNano: Long get() = delegate.startTime
```
```suggestion
    override val startTimeNanos: Long get() = delegate.startTime
```

0xnm commented on features/dd-sdk-android-trace/src/main/kotlin/com/datadog/android/trace/impl/internal/DatadogTraceIdAdapter.kt:
```diff
@@ -8,7 +8,7 @@ package com.datadog.android.trace.impl.internal
 import com.datadog.trace.api.DDTraceId
 
-internal class DatadogTraceIdAdapter(private val delegate: DDTraceId) : DatadogTraceId, DDTraceId() {
+internal data class DatadogTraceIdAdapter(private val delegate: DDTraceId) : DatadogTraceId, DDTraceId() {
```
do we need it to be `data` class?

satween replied:
`data` class force kotlin compiler generate `hashCode` and `equals` methods, that take into account every property declared in constructor, which required in tests, allowing to do such comparison: 

```kotlin 
assertThat(testedAgentSpanContextAdapter.traceId).isEqualTo(DatadogTraceIdAdapter(ddTraceId))
```

0xnm commented on features/dd-sdk-android-trace/src/test/kotlin/com/datadog/android/trace/impl/DatadogSpanAdapterTest.kt:
```diff
@@ -0,0 +1,229 @@
+    fun `M delegate setMetric(String, Int) W drop is called`() {
+        testedSpanAdapter.setMetric(fakeString, fakeInt)
+
+        verify(mockAgentSpan).setMetric(fakeString as CharSequence, fakeInt)
```
why this cast is needed? is there method overload ambiguity here?

satween replied:
debug leftover, removed

0xnm commented on features/dd-sdk-android-trace/src/test/kotlin/com/datadog/android/trace/impl/DatadogScopeAdapterTest.kt:
minor: it would be nice to add `Given-When-Then` in this files and also in other new test files as well.

satween replied:
✅

0xnm reviewed (commented):
I briefly did a review only for the commit on top of the top PR merge, will do a more detailed round tomorrow.

0xnm commented on features/dd-sdk-android-trace/src/test/kotlin/com/datadog/android/trace/impl/DatadogSpanBuilderAdapterTest.kt:
```diff
@@ -0,0 +1,158 @@
+
+    @Test
+    fun `M delegate withOrigin W withOrigin is called`() {
+        assertThat(testedBuilderAdapter.withOrigin(fakeString)).isEqualTo(testedBuilderAdapter)
```
minor: It is better to extract result of `testedBuilderAdapter.withOrigin(fakeString)` into a variable of `When` block. Same for other similar sites.

0xnm commented on features/dd-sdk-android-trace/src/test/kotlin/com/datadog/android/trace/impl/DatadogTracerAdapterTest.kt:
```diff
@@ -0,0 +1,141 @@
+
+    @BeforeEach
+    @Suppress("DEPRECATION")
+    fun `se tup`() {
```
```suggestion
    fun `set up`() {
```

0xnm commented on features/dd-sdk-android-trace/src/test/kotlin/com/datadog/android/trace/impl/DatadogTracerAdapterTest.kt:
```diff
@@ -0,0 +1,141 @@
+    private var fakeBool: Boolean = false
+
+    @BeforeEach
+    @Suppress("DEPRECATION")
```
what is deprecated here?

satween replied:
AgentTracer's `buildSpan(String)` method. Moved closer to class site

0xnm commented on features/dd-sdk-android-trace/src/test/kotlin/com/datadog/android/trace/impl/DatadogTracerAdapterTest.kt:
```diff
@@ -0,0 +1,141 @@
+    fun `M return wrapped span W activeSpan`() {
+        whenever(mockTracer.activeSpan()).thenReturn(mockSpan)
+
+        val span = testedTracer.activeSpan() as DatadogSpanAdapter
```
seems like the case here and below is redundant, since `assertThat` call below is just for the generic object, so `isEqualTo` should work without any cast as well.

If we want to check the type explicitly, we need to use `isInstanceOf`

0xnm commented on features/dd-sdk-android-trace/src/test/kotlin/com/datadog/android/trace/impl/DatadogTracerBuilderAdapterTest.kt:
```diff
@@ -0,0 +1,152 @@
+    private lateinit var testedBuilder: DatadogTracerBuilderAdapter
+
+    @Mock
+    lateinit var mockSdk: FeatureSdkCore
```
```suggestion
    lateinit var mockSdkCore: FeatureSdkCore
```

0xnm commented on features/dd-sdk-android-trace/src/test/kotlin/com/datadog/android/trace/impl/DatadogTracerBuilderAdapterTest.kt:
```diff
@@ -0,0 +1,152 @@
+
+    @Test
+    fun `M delegate CoreTracerBuilder#idGenerationStrategy W idGenerationStrategy`() {
+        testedBuilder.setTraceId128BitGenerationEnabled(fakeBoolean)
+
+        mockBuilder.idGenerationStrategy(IdGenerationStrategy.fromName("SECURE_RANDOM", fakeBoolean))
```
it seems we are not checking anything here? just doing some calls

0xnm commented on features/dd-sdk-android-trace/src/test/kotlin/com/datadog/android/trace/impl/DatadogTracingTest.kt:
```diff
@@ -0,0 +1,163 @@
+        val tracer = DatadogTracing.newTracerBuilder(mockSdkCore).build() as DatadogTracerAdapter
+
+        // THEN
+        assertThat(tracer).isNotNull
```
this is redundant, because we are already casting above to non-nullable type

0xnm commented on features/dd-sdk-android-trace/src/test/kotlin/com/datadog/android/trace/impl/DatadogTracingTest.kt:
```diff
@@ -0,0 +1,163 @@
+
+        // THEN
+        assertThat(tracer).isNotNull
+        val coreTracer: CoreTracer = tracer.delegate as CoreTracer
```
```suggestion
        val coreTracer = tracer.delegate as CoreTracer
```

0xnm commented on features/dd-sdk-android-trace/src/test/kotlin/com/datadog/android/trace/impl/DatadogTracingTest.kt:
```diff
@@ -0,0 +1,163 @@
+
+    @Test
+    fun `M log a maintainer error W build { TracingFeature not implementing InternalCoreTracerWriterProvider }`() {
+        // GIVEN
+
+        whenever(mockTracingFeature.unwrap<Feature>()) doReturn mock()
```
```suggestion
        // GIVEN
        whenever(mockTracingFeature.unwrap<Feature>()) doReturn mock()
```

0xnm commented on features/dd-sdk-android-trace/src/test/kotlin/com/datadog/android/utils/forge/DDTraceIdForgeryFactory.kt:
```diff
@@ -0,0 +1,16 @@
+ */
+package com.datadog.android.utils.forge
+
+import com.datadog.trace.api.DDTraceId
```
so now we have 2 API packages: one is from `dd-trace-java` with things like `DDTraceId` and another from us with things like `DatadogTraceId`?

0xnm commented on features/dd-sdk-android-trace/src/test/kotlin/com/datadog/android/utils/forge/DatadogSpanContextForgeryFactory.kt:
```diff
@@ -0,0 +1,21 @@
+import org.mockito.kotlin.mock
+
+class DatadogSpanContextForgeryFactory : ForgeryFactory<DatadogSpanContext> {
+    override fun getForgery(forge: Forge): DatadogSpanContext = mock {
+        on { traceId } doReturn forge.getForgery()
+        on { spanId } doReturn forge.aLong()
+        on { samplingPriority } doReturn forge.anInt()
+        on { tags } doReturn forge.aMap { aString() to aString() }
+    }
```
minor: maybe it is better to avoid mocking here and return a real object

satween replied:
mocking here allows not to bound to specific implementation of a `DatadogSpanContext ` let's keep it as it is

0xnm commented on instrumented/integration/src/main/kotlin/com/datadog/android/sdk/integration/RuntimeConfig.kt:
```diff
@@ -28,6 +29,8 @@ internal object RuntimeConfig {
     const val CONTENT_TYPE_TEXT = "text/plain;charset=UTF-8"
     private const val LOCALHOST = "http://localhost"
+    private const val SAMPLE_ALL = 100.0
+    private const val SEND_EACH_SPAN_THRESHOLD = 1
```
```suggestion
    private const val FLUSH_ON_EACH_SPAN_THRESHOLD = 1
```

0xnm commented on integrations/dd-sdk-android-okhttp-otel/src/main/kotlin/com/datadog/android/okhttp/otel/OkHttpExt.kt:
```diff
@@ -19,31 +17,13 @@ import okhttp3.Request
+    if (span !is OtelSpan) return@apply
+
     // very fragile and assumes that Datadog Tracer is used
     // we need to trigger sampling decision at this point, because we are doing context propagation out of OpenTelemetry
```
we don't need to trigger sampling decision anymore or we just now can do it at the OkHttp layer?

satween replied:
as discussed before - reverted to previous logic

0xnm commented on integrations/dd-sdk-android-okhttp-otel/src/test/kotlin/com/datadog/android/okhttp/otel/OkHttpExtTest.kt:
```diff
@@ -60,21 +62,31 @@ internal class OkHttpExtTest {
         }
-        expectedPrioritySampling =
-            if (fakeIsSampled) PrioritySampling.USER_KEEP else PrioritySampling.UNSET
+        expectedPrioritySampling = if (fakeIsSampled) {
+            DatadogTracingConstants.PrioritySampling.USER_KEEP
+        } else {
+            DatadogTracingConstants.PrioritySampling.UNSET
+        }
```
seems like we don't need it anymore?

0xnm commented on integrations/dd-sdk-android-sqldelight/src/main/kotlin/com/datadog/android/sqldelight/SqlDelightExt.kt:
```diff
@@ -30,7 +29,7 @@ fun <T : Transacter> T.transactionTraced(
     body: TransactionWithSpanAndWithoutReturn.() -> Unit
 ) {
-    withinSpan(operationName, GlobalTracer.get().activeSpan()) {
+    withinSpan(operationName, GlobalDatadogTracerHolder.get().activeSpan()) {
```
maybe we can simply call it `GlobalDatadogTracer` instead of `GlobalDatadogTracerHolder`? wdyt?

0xnm commented on reliability/single-fit/okhttp/src/test/kotlin/com/datadog/android/okhttp/tests/assertj/SpansPayloadAssert.kt:
```diff
@@ -72,15 +72,16 @@ internal class SpansPayloadAssert(actual: JsonObject) :
             assertThat(actualSpanId).overridingErrorMessage(
                 "Expected spanId to be $spanId but was $actualSpanId for index $index"
-            ).isEqualTo(spanId)
+            )
+                .isEqualTo(spanId)
```
minor: seems like redundant formatting change

0xnm commented on reliability/single-fit/trace/src/test/kotlin/com/datadog/android/trace/integration/otel/TracerExt.kt:
```diff
@@ -6,21 +6,23 @@
+private const val WRITER_FIELD_NAME = "coreTracerDataWriter"
+
+internal fun FeatureScope?.useBlockingWriter(): BlockingWriterWrapper {
+    val feature = this?.unwrap<Feature>() ?: throw NullPointerException("Can't cast to Feature")
```
minor: normally cast to `Feature` should always succeed, because the method is declared as `fun <T : Feature> unwrap(): T`. If it doesn't succeed, it will throw `ClassCastException`.

the only way to get `null` here is to have `null` receiver.

0xnm commented on sample/benchmark/src/main/java/com/datadog/benchmark/sample/ui/trace/TraceScenarioViewModel.kt:
```diff
@@ -182,14 +182,14 @@ internal class TraceScenarioViewModel(
         return viewModelScope.launch(defaultDispatcher) {
             val rootSpan = tracer.spanBuilder(task.config.spanOperation).apply {
-                setAttribute(DDTags.RESOURCE_NAME, task.config.spanResource)
+                setAttribute(DatadogTracingConstants.DDTags.RESOURCE_NAME, task.config.spanResource)
```
by looking on such usage, I'm wondering if we should rename `DatadogTracingConstants.DDTags` -> `DatadogTracingConstants.Tags`. wdyt? or maybe even `DatadogTracing.Tags`

satween replied:
The main idea for such naming was to reduce amount of potential mistyping during migration and make PR reviewing easier by keeping prefixes the same. 
Indeed, there is no reason to keep them separately - moved into tags ✅

0xnm reviewed (changes requested):
Big work!

I've added some comments, but overall looks good.

We also have some comments from late review here https://github.com/DataDog/dd-sdk-android/pull/2783#discussion_r2197840640 to address.

satween committed:
RUM-9902: post review fixes

satween commented on features/dd-sdk-android-trace/src/main/kotlin/com/datadog/android/trace/GlobalDatadogTracerHolder.kt:
```diff
@@ -0,0 +1,53 @@
+object GlobalDatadogTracerHolder {
+
+    @Volatile
+    @get:Synchronized
```
always forgetting this rule 👍

satween commented on features/dd-sdk-android-trace/api/apiSurface:
```diff
@@ -29,9 +19,32 @@ data class com.datadog.android.trace.TraceConfiguration
+  fun fromHex(String): com.datadog.android.trace.api.trace.DatadogTraceId
+  fun toLong(com.datadog.android.trace.api.trace.DatadogTraceId): Long
+  fun toHexString(com.datadog.android.trace.api.trace.DatadogTraceId): String
+object com.datadog.android.trace.impl.internal.DatadogTracingInternalToolkit
```
done ✅

satween committed:
RUM-9902: post review fixes

jonathanmos commented on features/dd-sdk-android-trace-api/src/main/kotlin/com/datadog/android/trace/api/DatadogTracingConstants.kt:
```diff
@@ -0,0 +1,217 @@
+        /**
+         *  String representing the error message.
+         */
+        const val ERROR_MSG: String = "error.message"
```
I see that we have both `ERROR_MSG` and `KEY_ERROR_MSG`, and also `ERROR_TYPE` and `KEY_ERROR_TYPE` with the same values. Is this on purpose?

satween replied:
ERROR_MSG - it's part of dd-trace-java. 
KEY_ERROR_MSG -part of public API for internal trace layer. 

Yes, it'n on purpose, I want to keep public only those constants that is being used by other modules in the SDK,

satween replied:
thanks for this catch, yep, was renamed during one of review fixes. Aligned now

jonathanmos commented on features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/telemetry/internal/TelemetryEventHandler.kt:
```diff
@@ -440,16 +440,17 @@ internal class TelemetryEventHandler(
         // for obfuscation enabled case.
         return try {
-            val globalTracerClass = Class.forName("io.opentracing.util.GlobalTracer")
+            val globalDatadogTracerHolderClass =
```
I don't see this GlobalDatadogTracerHolder class anywhere in the project - was it renamed?

jonathanmos reviewed (commented):
Nice pr! I've left a few comments

satween commented on integrations/dd-sdk-android-okhttp/src/test/kotlin/com/datadog/android/okhttp/trace/TracingInterceptorContextInjectionSampledTest.kt:
```diff
@@ -1027,7 +1029,7 @@ internal class TracingInterceptorContextInjectionSampledTest {
         verify(mockSpan).setTag("error.msg", throwable.message)
         verify(mockSpan).setTag("error.stack", throwable.loggableStackTrace())
-        verify(mockSpan).setTag(Tags.SPAN_KIND, Tags.SPAN_KIND_CLIENT)
+        verify(mockSpan).setTag("span.kind", "client")
```
in a previous implementation (OpenTracing) Tags was not only string, but also a `StringTag` object (which basically wraps the string). With opentracing removal this tests also changed to simple strings

satween committed:
RUM-9902: post review fixes

satween committed:
Merge pull request #2786 from DataDog/tvaleev/feature/RUM-9903

[RUM 9903] Updating opentelemetry version

satween committed:
RUM-9902: post review fixes

0xnm commented on integrations/dd-sdk-android-sqldelight/src/main/kotlin/com/datadog/android/sqldelight/internal/SqlDelightInternalExt.kt:
```diff
@@ -6,31 +6,30 @@
     } finally {
         span.finish()
-        scope.close()
+        scope?.close()
```
so now return type of `activateSpan` is nullable?

0xnm commented on reliability/single-fit/okhttp/src/test/kotlin/com/datadog/android/okhttp/HeadBasedSamplingTest.kt:
```diff
@@ -81,6 +79,12 @@ class HeadBasedSamplingTest {
     }
 
+    private fun Request.Builder.parentSpan(span: DatadogSpan): Request.Builder {
+        @Suppress("UnsafeThirdPartyFunctionCall") // Span can't be null
```
normally we don't have this check running against this module, so this `Suppress` is not really needed

0xnm commented on reliability/single-fit/okhttp/src/test/kotlin/com/datadog/android/okhttp/HeadBasedSamplingTest.kt:
```diff
@@ -341,24 +345,20 @@ class HeadBasedSamplingTest {
-                hasSamplingPriority(PrioritySampling.SAMPLER_KEEP)
+                hasParentId("0000000000000000")
+                hasRulePsr(1.0)
+                hasSamplingPriority(DatadogTracingConstants.PrioritySampling.USER_KEEP)
```
why we now are having `USER_KEEP` instead of `SAMPLER_KEEP`?

satween replied:
Because CoreTracer operates slightly different from AndroidTracer.

0xnm commented on reliability/single-fit/okhttp/src/test/kotlin/com/datadog/android/okhttp/HeadBasedSamplingTest.kt:
```diff
@@ -368,16 +368,16 @@ class HeadBasedSamplingTest {
                 hasResource("http://${mockServer.hostName}:${mockServer.port}/")
                 hasNoAgentPsr()
-                hasNoSamplingPriority()
+                hasSamplingPriority(2)
```
is it a right change? also test name is `parent context = OpenTracing Span, parent sampled`, but there is no `OpenTracing` anymore, so maybe the test is wrong?

satween replied:
it's all related to USER_KEEP instead of SAMPLER_KEEP that CoreTracer implementation has instead of that one we had with OpenTracing.

0xnm commented on reliability/single-fit/trace/src/test/kotlin/com/datadog/android/trace/integration/opentracing/UtilitiesTest.kt:
```diff
@@ -62,7 +63,8 @@ class UtilitiesTest {
     @AfterEach
     fun `tear down`() {
-        GlobalTracer::class.java.setStaticValue("isRegistered", false)
+        DatadogTracingToolkit.clear()
```
I don't see such method at the `DatadogTracingToolkit` declaration. Is it extension method? upd: ok, seems some test extensions

Also the package for this test class is `com.datadog.android.trace.integration.opentracing` - should we update it/reorganize?

satween replied:
1. Yes, moved all tests-related code into fixtures, so we could avoid reflection as much as possible 
2. Good notice, renamed

0xnm commented on reliability/single-fit/trace/src/test/kotlin/com/datadog/android/trace/integration/otel/OtelTraceConfigurationTest.kt:
```diff
@@ -49,31 +50,32 @@ class OtelTraceConfigurationTest {
         stubSdkCore = StubSDKCore(forge)
     }
 
-    @RepeatedTest(10)
+    @Test
```
is there a need for such change?

0xnm reviewed (commented):
I left some questions/comments, but overall lgtm.

satween committed:
Revert "RUM-9902 - fix tests, getting rid from TraceContext"

This reverts commit 4cfff410

# Conflicts:
#	integrations/dd-sdk-android-okhttp-otel/src/main/kotlin/com/datadog/android/okhttp/otel/OkHttpExt.kt
#	integrations/dd-sdk-android-okhttp-otel/src/test/kotlin/com/datadog/android/okhttp/otel/OkHttpExtTest.kt

satween commented on features/dd-sdk-android-trace-api/src/main/kotlin/com/datadog/android/trace/api/trace/DatadogTraceId.kt:
```diff
@@ -0,0 +1,39 @@
+     *
+     * @return the numeric value of the trace ID as a `Long`.
+     */
+    fun toLong(): Long
+
+    companion object
+}
```
I keep empty companion object here to simulate static methods: https://github.com/DataDog/dd-sdk-android/pull/2773#discussion_r2201441829

satween commented on features/dd-sdk-android-trace-api/src/main/kotlin/com/datadog/android/trace/api/scope/DataScopeListener.java:
```diff
@@ -4,13 +4,12 @@
 
 /** Hooks for scope activation */
-public interface ScopeListener {
+public interface DataScopeListener {
```
1. To keep it consistent with other classes declared in api. 
2. To being able to distinguish `-api` interface from `-trace-internal` interface

0xnm replied:
> To keep it consistent with other classes declared in api.

But other classes have `Datadog` prefix, not `Data`.

satween committed:
RUM-9902 - post review fixes

satween committed:
RUM-9902 - post review fixes

satween committed:
Apply suggestions from code review

Co-authored-by: Nikita Ogorodnikov <4046447+0xnm@users.noreply.github.com>

satween committed:
Apply suggestions from code review

Co-authored-by: Nikita Ogorodnikov <4046447+0xnm@users.noreply.github.com>

satween commented on features/dd-sdk-android-trace-otel/src/main/java/com/datadog/opentelemetry/trace/OtelExtractedContext.java:
```diff
@@ -57,33 +60,37 @@ public static AgentSpan.Context extract(Context context, InternalLogger logger)
     @Override
-    public int getSamplingPriority() {
-        return this.prioritySampling;
+    public Map<String, Object> getTags() {
+        // Do nothing
+        return Collections.emptyMap();
     }
```
Only if we want to support some specific case

satween committed:
RUM-9902 - post review fixes

satween committed:
RUM-9902 - post review fixes

satween committed:
RUM-9902 - post review fixes

satween committed:
Merge remote-tracking branch 'origin/tvaleev/feature/RUM-9902-removing-unused-code' into tvaleev/feature/RUM-9902-removing-unused-code

satween committed:
RUM-9902 - post review fixes

satween commented on features/dd-sdk-android-trace-otel/src/main/kotlin/com/datadog/android/trace/opentelemetry/OtelTracerProvider.kt:
```diff
@@ -360,27 +294,22 @@ class OtelTracerProvider internal constructor(
             "You are trying to bundle the traces with a RUM context, " +
                 "but the RUM context is missing. " +
                 "You should check if the RUM feature is enabled for your SDK instance."
+        internal const val USED_AS_GLOBAL_INFO_MESSAGE =
+            "OpenTracer's DatadogTracer instance will be used as global instance"
+        internal const val NOT_USED_AS_GLOBAL_INFO_MESSAGE =
+            "OpenTracer's DatadogTracer instance will not be used as global instance"
```
`OtelTracer` *

satween commented on features/dd-sdk-android-trace-otel/src/main/kotlin/com/datadog/android/trace/opentelemetry/OtelTracerProvider.kt:
```diff
@@ -144,60 +133,46 @@ class OtelTracerProvider internal constructor(
+                .withServiceName(serviceName)
+                .build()
+
+            GlobalDatadogTracer.registerIfAbsent(datadogTracer).let { registeredAsGlobal ->
+                sdkCore.internalLogger.log(
+                    InternalLogger.Level.INFO,
+                    listOf(
+                        InternalLogger.Target.USER,
+                        InternalLogger.Target.MAINTAINER
+                    ),
+                    { if (registeredAsGlobal) USED_AS_GLOBAL_INFO_MESSAGE else NOT_USED_AS_GLOBAL_INFO_MESSAGE }
+                )
             }
```
I was thinking to add it automatically in order to make configuration simplier, but, I gues, it's better to make such changes in a separate brunch

satween commented on instrumented/integration/src/androidTest/kotlin/com/datadog/android/sdk/integration/trace/TracesTest.kt:
```diff
@@ -151,7 +171,7 @@ internal abstract class TracesTest {
         const val META_KEY = "meta"
         const val METRICS_KEY = "metrics"
-        internal val INITIAL_WAIT_MS = TimeUnit.SECONDS.toMillis(60)
+        internal val INITIAL_WAIT_MS = TimeUnit.SECONDS.toMillis(100)
```
I've caught timeout few times with this tests, let's revert and see if it passes

satween committed:
Update integrations/dd-sdk-android-okhttp/src/main/kotlin/com/datadog/android/okhttp/trace/TracingInterceptor.kt

Co-authored-by: Nikita Ogorodnikov <4046447+0xnm@users.noreply.github.com>

satween commented on integrations/dd-sdk-android-okhttp/src/test/kotlin/com/datadog/android/okhttp/internal/utils/SpanSamplingIdProviderTest.kt:
```diff
@@ -38,17 +40,17 @@ import org.mockito.quality.Strictness
+    lateinit var mockSpanContext: DatadogSpanContext
 
-    lateinit var fakeTags: Map<String, Any>
+    private lateinit var fakeTags: Map<String?, Any?>
```
This is how AS converts it from java, fixed

satween commented on features/dd-sdk-android-trace/src/main/kotlin/com/datadog/android/trace/GlobalDatadogTracer.kt:
```diff
@@ -0,0 +1,63 @@
+ */
+object GlobalDatadogTracer {
+
+    @get:Synchronized
```
Replaced with more thread-safe approach

satween committed:
Merge remote-tracking branch 'origin/tvaleev/feature/RUM-9902-removing-unused-code' into tvaleev/feature/RUM-9902-removing-unused-code

0xnm commented on reliability/single-fit/trace/src/test/kotlin/com/datadog/android/trace/integration/api/UtilitiesTest.kt:
```diff
@@ -167,13 +173,13 @@ class UtilitiesTest {
 
     @RepeatedTest(16)
-    fun `M send span with exception W buildSpan() + start() + logThrowable() + finish()`(
+    fun `M send span with exception W buildSpan() + start() + logThrowable () + finish()`(
```
```suggestion
    fun `M send span with exception W buildSpan() + start() + logThrowable() + finish()`(
```

satween commented on features/dd-sdk-android-trace/src/main/kotlin/com/datadog/android/trace/GlobalDatadogTracer.kt:
```diff
@@ -0,0 +1,59 @@
+ */
+object GlobalDatadogTracer {
+
+    private val instance = AtomicReference<DatadogTracer?>()
```
yes, because getters are not synchronized

satween committed:
RUM-9902 - post review fixes

satween commented on features/dd-sdk-android-trace/src/main/kotlin/com/datadog/android/trace/internal/DatadogTracingToolkit.kt:
```diff
@@ -0,0 +1,70 @@
+ * Provides implementation for specific interfaces to dependent modules
+ */
+@InternalApi
+object DatadogTracingToolkit {
```
This object is a bunch of functions basically, the state provided only for tests. 
propagationHelper/spanIdConverter also just proxy to static functions, just aggregated logically

satween commented on features/dd-sdk-android-trace/src/main/kotlin/com/datadog/android/trace/internal/DatadogTracingToolkit.kt:
```diff
@@ -0,0 +1,70 @@
+     * Provides an instance of [DatadogSpanLogger] for logging span-related messages, errors,
+     * and attributes. Selects the appropriate logger implementation based on the available context.
+     */
+    var spanLogger: DatadogSpanLogger
+        internal set(value) {
+            spanLoggerNullable = value
+        }
+        get() = spanLoggerNullable
+            ?: (Datadog.getInstance() as? FeatureSdkCore)?.let(::DatadogSpanLoggerAdapter)
```
good catch - refactored

satween commented on features/dd-sdk-android-trace-api/src/main/kotlin/com/datadog/android/trace/api/tracer/DatadogTracerBuilder.kt:
```diff
@@ -0,0 +1,83 @@
+     * is provided this property and its related logic will be ignored.
+     * @param traceRateLimit the trace rate limit as a value between 1 and Int.MAX_VALUE (default is Int.MAX_VALUE)
+     */
+    fun withTraceRateLimit(@IntRange(from = 1, to = Int.MAX_VALUE.toLong()) traceRateLimit: Int): DatadogTracerBuilder
```
This comes from Otel and supported by CoreTracer,  but I am okay to make it internal for now

0xnm replied:
Let's keep it then, apparently it is already a part of `OtelTracerProvider` we have right now in our codebase

satween commented on features/dd-sdk-android-trace/src/main/kotlin/com/datadog/android/trace/internal/DatadogPropagationHelper.kt:
```diff
@@ -0,0 +1,53 @@
+     * @return True if the context is identified as an extracted context, otherwise false.
+     */
+    fun isExtractedContext(context: DatadogSpanContext): Boolean {
+        if (context !is DatadogSpanContextAdapter) return false
```
not sure we should. Otherwise we should log everywhere when it's failed to cast to adapter. WDYT?

0xnm commented on features/dd-sdk-android-trace/src/main/kotlin/com/datadog/android/trace/internal/DatadogTracingToolkit.kt:
```diff
@@ -0,0 +1,68 @@
+    var propagationHelper: DatadogPropagationHelper = DatadogPropagationHelper()
+        internal set
+
+    internal var builderProvider: DatadogTracerBuilder? = null
```
let's call it `testBuilderProvider`? it is set only in tests in seems.

I'm not quite a fan of having test references in the production code, but on the opposite side is mocking static `newTracerBuilder` - maybe we can create a TODO item to somehow improve such setup in the future.

0xnm commented on features/dd-sdk-android-trace/src/main/kotlin/com/datadog/android/trace/internal/DatadogTracingToolkit.kt:
```diff
@@ -0,0 +1,68 @@
+     * @param builder the tracer builder to set the trace rate limit for
+     * @param traceRateLimit the trace rate limit as a value between 1 and Int.MAX_VALUE (default is Int.MAX_VALUE)
+     */
+    fun setTraceRateLimit(
```
let's put it back, see my comment https://github.com/DataDog/dd-sdk-android/pull/2773#discussion_r2215224420

0xnm reviewed (approved):
LGTM. I left two comments, one of them, regarding `traceRateLimit`, should be addressed before merging this PR.

Also I have a concern about `SAMPLE_KEEP` -> `USER_KEEP` change for the sampling priority. We need to check with APM spec if it may have any impact on the intake processing (e.g. metrics computation) or trace propagation.

Maybe it has impact only for the tracer -> agent -> intake setup, and since we are in the agent-less mode and setup is tracer -> intake, this doesn't change anything for us, but still it is better to double-check.

This can be done after merging PR.
